pkg/entrypoint/run.go

```diff
@@ -0,0 +1,295 @@
+func (o Options) ExecuteProcess() (int, error) {
+	if o.ArtifactDir != "" {
+		if err := os.MkdirAll(o.ArtifactDir, os.ModePerm); err != nil {
+			return InternalErrorCode, fmt.Errorf("could not create artifact directory(%s): %v", o.ArtifactDir, err)
+		}
+	}
+	processLogFile, err := os.Create(o.ProcessLog)
+	if err != nil {
+		return InternalErrorCode, fmt.Errorf("could not create process logfile(%s): %v", o.ProcessLog, err)
+	}
+	defer processLogFile.Close()
+
+	output := io.MultiWriter(os.Stdout, processLogFile)
+	logrus.SetOutput(output)
+	defer logrus.SetOutput(os.Stdout)
+
+	// ---
+	timeout := optionOrDefault(o.Timeout, DefaultTimeout)
+	gracePeriod := optionOrDefault(o.GracePeriod, DefaultGracePeriod)
+	var commandErr error
+	cancelled, aborted := false, false
+
+	done := make(chan error)
+	go func() {
+		done <- o.waitForPrevStep()
+	}()
+	select {
+	case err := <-done:
+		commandErr = err
+		if err != nil {
+			cancelled = true
+		}
+	case <-time.After(timeout):
+		logrus.Errorf("Previous step did not finish within %s timeout", timeout)
+		cancelled = true
+	}
+
+	var returnCode int
+	if cancelled {
+		if aborted {
+			commandErr = errAborted
+			returnCode = AbortedErrorCode
+		} else {
+			commandErr = errTimedOut
+			returnCode = InternalErrorCode
+		}
+		return returnCode, commandErr
+	}
+
+	// ---
+	executable := o.Args[0]
+	var arguments []string
+	if len(o.Args) > 1 {
+		arguments = o.Args[1:]
+	}
+	command := exec.Command(executable, arguments...)
+	command.Stderr = output
+	command.Stdout = output
+	if err := command.Start(); err != nil {
+		return InternalErrorCode, fmt.Errorf("could not start the process: %v", err)
+	}
+
+	// if we get asked to terminate we need to forward
+	// that to the wrapped process as if it timed out
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt, syscall.SIGTERM)
+
+	done = make(chan error)
+	go func() {
+		done <- command.Wait()
+	}()
+	select {
+	case err := <-done:
+		commandErr = err
+		if o.ShouldRunPostRun {
+			o.postRunWriteFile(0)
+		}
+
+	case <-time.After(timeout):
+		logrus.Errorf("Process did not finish before %s timeout", timeout)
+		cancelled = true
+		gracefullyTerminate(command, done, gracePeriod)
+	case s := <-interrupt:
+		logrus.Errorf("Entrypoint received interrupt: %v", s)
+		cancelled = true
+		aborted = true
+		gracefullyTerminate(command, done, gracePeriod)
+	}
+
+	if cancelled {
+		if aborted {
+			commandErr = errAborted
+			returnCode = AbortedErrorCode
+		} else {
+			commandErr = errTimedOut
+			returnCode = InternalErrorCode
+		}
+	} else {
+		if status, ok := command.ProcessState.Sys().(syscall.WaitStatus); ok {
+			returnCode = status.ExitStatus()
+		} else if commandErr == nil {
+			returnCode = 0
+		} else {
+			returnCode = 1
+		}
+
+		if returnCode != 0 {
+			commandErr = fmt.Errorf("wrapped process failed: %v", commandErr)
+		}
+	}
+	return returnCode, commandErr
+}
+
+func (o *Options) mark(exitCode int) error {
```

Comment:
what does `mark` mean? can we be a bit more specifica bout what this function does?